[lldb] Auto-detect whether to use the static bindings

JDevlieghere opened the pull request:
Use LLDB's auto-detection logic to determine whether the developer has a supported version of SWIG installed. Only use the static bindings if that's not the case, or the developer has explicitly set the LLDB_USE_STATIC_BINDINGS flag.

JDevlieghere committed:
[lldb] Auto-detect whether to use the static bindings

Use LLDB's auto-detection logic to determine whether the developer has a
supported version of SWIG installed. Only use the static bindings if
that's not the case, or the developer has explicitly set the
LLDB_USE_STATIC_BINDINGS flag.

JDevlieghere commented:
@swift-ci please test

bulbazord commented on lldb/cmake/modules/LLDBConfig.cmake:
```diff
@@ -56,6 +56,19 @@ macro(add_optional_dependency variable description package found)
 endmacro()
 
+# BEGIN SWIFT MOD
+# Auto-detect whether we should use the static bindings.
+# This needs to happen before we detect any of the other dependencies because
+# their behavior relies on whether LLDB_USE_STATIC_BINDINGS is enabled or not.
+add_optional_dependency(LLDB_ENABLE_SWIG "Enable using SWIG to generate the LLDB bindings" SWIG SWIG_FOUND VERSION 3)
```
I think this could probably be put on llvm.org. SWIG is technically an optional dependency since `LLDB_ENABLE_PYTHON` can be set to `OFF`. Same for Lua. Shouldn't be a blocker though.

JDevlieghere replied:
As is, I don't think it would be meaningful to put this upstream. There are no static bindings there. We could change the way we autodetect Python and Lua support upstream though. Currently, we have `FindPythonAndSwig` and `FindLuaAndSwig` which only succeed if we find both. We could hoist finding SWIG out of those files, but we'd still have to check `LLDB_ENABLE_SWIG`. Is that what you're suggesting?

bulbazord replied:
Yes, something along those lines. I agree that this is not meaningful to put upstream in its current form. Upstream could be changed to support a `LLDB_ENABLE_SWIG` in the way you described, but I think that would be non-trivial work.

JDevlieghere replied:
https://reviews.llvm.org/D138879

bulbazord reviewed (approved):
LGTM, one nit but not a blocker.

JDevlieghere committed:
[lldb] Warn when static bindings are out of sync

Print a warning when the static bindings are out of sync. The warning
only triggers when not using the static bindings (because otherwise
they'll be identical by design). The warning also includes the command
to update them in the project source.

JDevlieghere committed:
[lldb] Make SWIG an auto-detected dependency

This patch makes SWIG itself an auto-detected dependency. This allows us
to look for SWIG once in a centralized place and makes it easier
downstream to detect whether to use the static bindings.

Differential revision: https://reviews.llvm.org/D138879

JDevlieghere commented:
@swift-ci please test

JDevlieghere commented:
@swift-ci please test
